Add benchTradeStats()

vi-to opened the pull request:
The present PR adds `benchTradeStats()` as a function that makes `benchTradePerf()` output available for statistical testing. Tests are touched on in #54.
 
With respect to their relation, `benchTradeStats()` is implemented as a new function as opposed to a `benchTradePerf()` method, in order to free the latter from unnecessary further complexity and burdensome structures. Nonetheless, because `benchTradePerf()` outputs are needed in all the prospected `benchTradeStats()` Porfolio-Symbol combinations, they are added to the former function output so that to avoid separately going through benchmarking under potentially heavy computations. 
This being said, `benchTradeStats()` flow itself may appear wider than its current scopes, this is done in the hope of allowing easier tests additions in the future.   

Most of the statistical tests included are called from the `stats` R-core package (https://stat.ethz.ch/R-manual/R-devel/library/stats/html/00Index.html).
The output from these tests is larger than their default calls and is stored almost as is, in order to give them complete authority on tests conditions evaluations and allow further usability. The `conf.level` and `alternative` params provide control of these tests, if possible I look forward to simplify this passing through ellipses. An exception to tests from `stats` is given by the 'Median' test under the independent approach, which is implemented as Kissell suggests (see `@references`).

As a remark to avoid confusion, I shall notice that we are conceptually overloading `blotter` objects dominant meaning, simply in order to take advantage of their data structures: 
- with respect to the portfolio object: from a theoretical standpoint we mean 'orders';
- with respect to the Symbol in the portfolio object: perhaps because of its pre-existent "accounting goals", the current structure of the object has well motivated strict rules to respect in the initialization and updating. Now, considering the "paired observations testing approach" we are really referring to the same Symbol, but we are comparing potentially different execution strategies (or brokers) on it. Therefore, a way to account for the need of this distinction using the existent structures is simply to create fictitious different symbols by giving them arbitrarily different names (e.g., we want to compare the execution of two brokers, broker A and broker B, on the same symbol and hence we initialize `ABC.A` and `ABC.B` for each `Portfolio`).    

The `account` object was not used here because it contains summaries of portfolios, to serve other purposes. From the docs: "An Account object is a list of portfolios with portfolio summary information [...] is modeled on the CFTC Thirteen-column presentation table". I would not be surprised if a more extensive use of an 'Order' notion in the package would motivate the need of a new specific object, perhaps based on the two existing ones.

vi-to committed:
Add benchTradeStats()

Introduce the new function for statistical testing of benchmarked transaction.
Extensive documentation is provided.

@importFrom adds statistical testing functions and a distribution function from `stats` package to the NAMESPACE

vi-to committed:
Provide more informative dialog statements and variable console names

The modifications in warning() should permit the user grasp which portfolio has to be modified, if any, depending on his objects initializations and analysis intentions.

The fix in testout$data.name is to avoid confusion with the rest of the "htest" object output, furthermore the variable 'nsuccesses' must clearly be nonnegative

vi-to committed:
Add an important precisation on what is the data tested

Also, adding a more informative summary table of benchmarked orders, which should support the interpretation on overall orders performance

vi-to committed:
Rename some internal and output variables

Trying to have meaningful distinct names on both sides of the function. Style camelCase avoided on 'benchout', 'benchtotest' and 'testout' to suggest they are all parts of 'benchTestOut' return. Not sure this is good or accepted practice

vi-to committed:
Simplify two lines of code in approach='independent'

Results from previous commits are unchanged. Changes are introduced to slightly improve code readability

vi-to committed:
Add DGP tests: Chi-Square and Kolmogorov-Smirnov GoF tests

The former is implemented as Kissell suggests (pp. 123--124), the latter is called from `stats` package. NAMESPACE and docs are updated accordingly. Also, new examples are added in docs.

One aspect which makes the function flow perhaphs less smooth and intuitive than needed is that for these tests Kissell suggests to use the arrival cost (a cost metric) as the preferred benchmark. This is in accordance to what suggested for the 'independent' approach, but in contrast suggestions for the 'paired' approach in which case performance metrics are considered (specifically, the VWAP).
There are clearly other reasons to keep the two approaches separate, but fix a common metric throughout may simplify some parts of the function; also taking into account that we deliberately kept out cost metrics from benchTradePerf().

Lastly, I shall stress some important details with respect to dgptest='ChiSq'. More specifically to how buckets are built. The way Kissell seems to suggest to use the standard-deviation based buckets building, namely {(-Inf,-3)*sd, [-3,-2.5]*sd, ..., [2.5, 3]*sd, (3, Inf)*sd}, is clearly not quite machine-friendly as it is. We could go through some fine tuning, but these considerations would neverthless be subordined to how findInterval() interprets numbers on close to the bounds and may quickly lead us to as deeply interesting as perhaps out-of-scope considerations. Although one may argue that chiBinStats will face a slight left-skewed bias, the test is ultimately based on their sum and thus results will be unbiased.
Therefore, to forgo cumbersome and perhpas unnecessary findInterval() splittings, the buckets have been litterally implemented as {(-Inf,-3]*sd, (-3,-2.5]*sd, ..., (2.5, 3]*sd, (3, Inf]*sd}. However and as an endnote on notation, the very last interval is printed as (3, Inf)*sd to avoid writing a mathematical abomination: Inf will never be reached anyways.

vi-to committed:
Add S3 print method to benchTradeStats()

Function output as it is may quickly saturate one's console, therefore excluding not-main-purpose output from console printing

jaymon0703 commented on R/benchTradeStats.R:
```diff
@@ -0,0 +1,296 @@
+#' approach ultimately relies on analysts' specific research questions. In turn, 
+#' each of them critically depend on how transactional data was obtained in the 
+#' first place and the characteristics it has.
+#' First, the \emph{paired samples approach}, where: trades are two equal-length 
```
This may be confusing to a user and is strictly not accurate. Child orders between the 2 strategies can be any number and any length, as long as they form part of parent orders that were of the same length. In other words, strategy A could have 100 child orders where each child that traded was a smaller quantity, whereas strategy B could have 20 child orders, in which the strategy was more passive (ie. in the market longer) and the average quantity was greater, in order to ensure ultimate parent order volume on strategy A and strategy B are the same.

vi-to replied:
Of course it would be in general, but this does not appear to be the case for the paired testing approach to be applicable. Here each order has to be, or have been, split into equal length pairs of child orders executed during the exact same timeframe.

jaymon0703 replied:
Overall timeframe must be equal agreed, but there is no requirement that the children be equal length or equal quantities. This will be impossible to constrain in production, and is irrelevant. I just want to be careful about the language when referring to the constraints for this test.

vi-to replied:
Yes, after all we are considering the overall performance (or cost) metric of each child order only and therefore agree. However, I would recommend transactions happened under each of them to be very close in number, preferably approaching equality. Clearly, one may not guarantee such a circumstance in general. This leads to restrictions on the kinds of algorithmic trading strategies that can be suitably tested under this approach, namely the fact that it is mainly indicated as a testing framework for "static trade schedule algorithm(s)". These concepts are left out from function docs both because we are looking at the same references and because they may take rather long discussions. I have vignette discussion intentions in mind, where we can discuss them taking a wider perspective rather than strictly provide function usage in a general context. Your feedback on documentation distinction would be much appreciated

vi-to committed:
First attempt to solve Travis.CI build

\donttest{} replaces \dontrun{} in @examples wrapping

jaymon0703 commented on R/benchTradeStats.R:
```diff
@@ -0,0 +1,296 @@
+#'
+#' @details
+#' In the paired samples approach we seek to test different trading strategies 
+#' (or brokers) on the same symbol, whereas in the indepented samples approach this 
```
In the paired samples approach we seek to test different trading strategies (or brokers) on the same symbol *over the same timeframe*, ...

Mentioning that the test is specific for 2 strategies trading at the exact same time is key, and worth emphasising

vi-to replied:
Completely agree on it. I will blend the docs to stress this.

However, in the meanwhile please note that this is at least quickly stated in the line below your [review comment](https://github.com/braverock/blotter/pull/96#discussion_r299021465)

vi-to committed:
Merge branch '54_algotca' of git://github.com/braverock/blotter into 54/tca/benchTradeStats

This to account for a fix in Travis.CI build failures raised by a tedious WARNING. The last one originated from a deprecated function used in updatePosPL(), as called by updatePortf(). In turn this last one was part of benchTradePerf() examples script that was checked in benchTradeStats() too, where the WARNING and thus the Travis.CI build failure ultimately propagated

jaymon0703 reviewed (commented):
Some comments on the docs, reviewing the code and output now...

vi-to committed:
Refine the docs and fix two examples

Respectively to:
- provide some spefications that will better blend with vignette
- have examples consistency after a renamed input variable definition

vi-to committed:
Introduce 'metric' param

This parameter allows to select the desired benchmark metric setting in terms of PnL performance or cost. And from then on to conduct all analyses constistently with the chosen interpretation.

Docs contains additional explanations and examples are updated accordingly

vi-to committed:
Simplify benchTradeStats() structure

Thanks to commit e857c7b8d312d1c1863e838964e84811aec2d845 generalization it is now possible to shorten the function control flow.
As usual, the hope is to improve its readability and maintenability

vi-to commented on R/benchTradeStats.R:
```diff
@@ -222,18 +236,23 @@ benchTradeStats <- function(Portfolio,
+  metricFullName <- ifelse(metric == 1, 'Performance', 'Cost') # consistency with benchTradePerf() naming style
+  metricAbbrName <- ifelse(metric == 1, 'Perf', 'Cost')
+  
+  benchTestOut <- benchoutstore <- list()
```
`benchoutstore` replaces `benchout` because of a new mechanism introduced in the code. I forgot to reassign this variable on line 393 of e857c7b8d312d1c1863e838964e84811aec2d845, in return values. This is accounted for in last commit 161e3f59372077e8e96b872a55e35ec8b7bf5eaf, to reproduce e857c7b8d312d1c1863e838964e84811aec2d845 version please add this change manually.

vi-to committed:
Fix 'Median' statistical hypothesis test

Conlusions were reversed. Also took the occasion to generalize the 'report' comment code

jaymon0703 reviewed (approved):
Thanks @vi-to

vi-to committed:
Fix typo in 'Median' test report
